File: C++/543.cpp
__________________________________________________________________________________________________
sample 12 ms submission
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    Solution() {
        ios_base::sync_with_stdio(false);
        cin.tie(0);
    }
    int Max = 0;
    int diameterOfBinaryTree(TreeNode*root) {
        if(root == nullptr) return 0;
        helper(root);
        return Max-1;
    }
private:
    int helper(TreeNode*&root){
         if (root == nullptr) return 0;
        int left = 1 + helper(root->left);
        int right = 1 + helper(root->right);
        if (right + left -1 > Max) Max = left + right -1;
        return max(left, right);
    }
};
__________________________________________________________________________________________________
sample 19848 kb submission
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int dia(TreeNode * root, int * ans){
          if(root==NULL){
            return 0;
        }
        if(root->left==NULL&&root->right==NULL){
            return 1;
        }
        int lefty = dia(root->left,ans);
        int righty = dia(root->right,ans);
        if(*ans<lefty+righty){
            *ans=lefty+righty;
        }
        int max=lefty;
        if(max<righty)
            max=righty;
        return max+1;
    }
    
    int diameterOfBinaryTree(TreeNode* root) {
        int ans=0;
        dia(root,&ans); 
        return ans;
    }
};
__________________________________________________________________________________________________

File: Python3/543.py
__________________________________________________________________________________________________
sample 36 ms submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.ans = 0
        
        def depth(p):
            if not p: return 0
            left, right = depth(p.left), depth(p.right)
            self.ans = max(self.ans, left+right)
            return 1 + max(left, right)
            
        depth(root)
        return self.ans
__________________________________________________________________________________________________
sample 15344 kb submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def recursion(self, node):
        if node == None: return 0
        else:
            dl = self.recursion(node.left)
            dr = self.recursion(node.right)
            self.max = max(dl+dr+1, self.max)
            return max(dl,dr) + 1
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        if root == None: return 0
        self.max = 0
        self.recursion(root)
        return self.max-1        
__________________________________________________________________________________________________
